sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/TooManyParameters.cs

```diff
@@ -0,0 +1,95 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.VisualBasic;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class TooManyParameters : TooManyParametersBase<SyntaxKind, ParameterListSyntax>
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        protected override GeneratedCodeRecognizer GeneratedCodeRecognizer { get; } = VisualBasicGeneratedCodeRecognizer.Instance;
+        protected override SyntaxKind[] SyntaxKinds { get; } = new SyntaxKind[] { SyntaxKind.ParameterList };
+
+        protected override string UserFriendlyNameForNode(SyntaxNode node) => nodeToDeclarationName[node.Kind()];
+
+        protected override int CountParameters(ParameterListSyntax parameterList) => parameterList.Parameters.Count;
+
+        protected override bool CanBeChanged(SyntaxNode node, SemanticModel semanticModel)
+        {
+            if (!nodeToDeclarationName.ContainsKey(node.Kind()))
+            {
+                return false;
+            }
+            if ((node as SubNewStatementSyntax)?.ParameterList?.Parameters.Count is int parameterCount &&
+                parameterCount > Maximum &&
+                node.Parent is ConstructorBlockSyntax constructorBlock &&
+                ContainsMyBaseNewInvocation(constructorBlock, Maximum))
+            {
+                return false;
+            }
+            if (node.IsAnyKind(LambdaHeaders))
+            {
+                return true;
+            }
+            return VerifyCanBeChangedBySymbol(node, semanticModel);
+        }
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager,
+                isEnabledByDefault: false);
+
+        private static readonly ImmutableDictionary<SyntaxKind, string> nodeToDeclarationName = new Dictionary<SyntaxKind, string>
+        {
+            { SyntaxKind.SubNewStatement, "Constructor" },
+            { SyntaxKind.FunctionStatement, "Function" },
+            { SyntaxKind.SubStatement, "Sub" },
+            { SyntaxKind.DelegateFunctionStatement, "Delegate" },
+            { SyntaxKind.DelegateSubStatement, "Delegate" },
+            { SyntaxKind.SubLambdaHeader, "Lambda" },
+            { SyntaxKind.FunctionLambdaHeader, "Lambda" },
+        }.ToImmutableDictionary();
+
+        private static readonly SyntaxKind[] LambdaHeaders = new SyntaxKind[]
```

Comment:
You could simplify the array creation.
```suggestion
        private static readonly SyntaxKind[] LambdaHeaders = new[]
```